src/VisualStudio/Core/Def/Implementation/UnusedReferences/RemoveUnusedReferencesCommandHandler.cs

```diff
@@ -0,0 +1,319 @@
+            var menuCommandService = (IMenuCommandService)_serviceProvider.GetService(typeof(IMenuCommandService));
+            if (menuCommandService != null)
+            {
+                AddCommand(menuCommandService, ID.RoslynCommands.RemoveUnusedReferences, Guids.RoslynGroupId, OnRemoveUnusedReferencesForSelectedProject, OnRemoveUnusedReferencesForSelectedProjectStatus);
+            }
+
+            return;
+
+            // Local functions
+            static OleMenuCommand AddCommand(
+                IMenuCommandService menuCommandService,
+                int commandId,
+                Guid commandGroup,
+                EventHandler invokeHandler,
+                EventHandler beforeQueryStatus)
+            {
+                var commandIdWithGroupId = new CommandID(commandGroup, commandId);
+                var command = new OleMenuCommand(invokeHandler, delegate { }, beforeQueryStatus, commandIdWithGroupId);
+                menuCommandService.AddCommand(command);
+                return command;
+            }
+        }
+
+        private void OnRemoveUnusedReferencesForSelectedProjectStatus(object sender, EventArgs e)
+        {
+            var command = (OleMenuCommand)sender;
+
+            // Only show the "Remove Unused Reference" menu commands for CPS based managed projects.
+            var visible = TryGetSelectedProjectHierarchy(out var hierarchy) &&
+                hierarchy.IsCapabilityMatch("CPS") &&
+                hierarchy.IsCapabilityMatch(".NET") &&
+                _workspace.Options.GetOption(FeatureOnOffOptions.OfferRemoveUnusedReferences);
+            var enabled = false;
+
+            if (visible)
+            {
+                enabled = !IsBuildActive();
+            }
+
+            if (command.Visible != visible)
+            {
+                command.Visible = visible;
+            }
+            if (command.Enabled != enabled)
+            {
+                command.Enabled = enabled;
+            }
+        }
+
+        private void OnRemoveUnusedReferencesForSelectedProject(object sender, EventArgs args)
+        {
+            if (TryGetSelectedProjectHierarchy(out var hierarchy))
+            {
+                Project? project = null;
+                ImmutableArray<ReferenceUpdate> referenceUpdates = default;
+                _threadOperationExecutor.Execute(ServicesVSResources.Remove_Unused_References, ServicesVSResources.Analyzing_project_references, allowCancellation: true, showProgress: true, (operationContext) =>
+                {
+                    (project, referenceUpdates) = GetUnusedReferencesForProjectHierarchy(hierarchy, operationContext.UserCancellationToken);
+                });
+
+                if (project is null ||
+                    referenceUpdates.IsEmpty)
+                {
+                    MessageDialog.Show(ServicesVSResources.Remove_Unused_References, ServicesVSResources.No_unused_references_were_found, MessageDialogCommandSet.Ok);
```

Comment:
Which MessageDialog is this? Will this properly put VS into whatever modal state it needs to be? The VS-aware APIs will do the right thing; if this is just Winforms/WPF I don't remember.

Reply:
This is the VS dialog.